### ISSUE TITLE
DataParallelModel seems to be stuck

### ISSUE THREAD

**Issue by jianchao-li** [zhanghang1989/PyTorch-Encoding#121]
Hello, @zhanghang1989.

I am working with your code to train a PSPNet50 on Pascal VOC. However, the training process was stuck at [this line](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/experiments/segmentation/train.py#L108). After a closer inspection, I found it was stuck at [`layer3` in `base_forward`](https://github.com/zhanghang1989/PyTorch-Encoding/blob/d4e195538a5cac3a33b9cbde28d54c8ed78a3f91/encoding/models/base.py#L59). The following is my training script.

```bash
#!/bin/bash

CUDA_VISIBLE_DEVICES=0,3 python train.py \
  --dataset pascal_voc \
  --model psp \
  --backbone resnet50 \
  --batch-size 8 \
  --aux
```

I am still looking into this problem. Since I found no similar issues, I posted it here. If you have any idea about this issue, could you please kindly share it with me? Thank you.

**Comment by zhanghang1989** [comment]
I don't have that problem on my machine. Please keep me posted how it goes.

**Comment by jianchao-li** [comment]
Hello, @zhanghang1989. Thank you for your reply. Now I switched to training using a single GPU with batch size 4. I am still reading your codes to understand what is happing. So I can only guess whether the problem could be caused by using two GPUs with incontiguous IDs, like `CUDA_VISIBLE_DEVICES=0,3`.

**Comment by jianchao-li** [comment]
Hello, @zhanghang1989. I noticed that you mentioned the following phenomenon in Zhihu.

```
3。本身前两步搞定之后，应该一切顺利了，可是训练的时候总是跑几十个iteration就卡死。这个其实是pytorch autograd engine 的问题， 因为每个BN layer的均值和方差都是cross gpu 的grad graph，而我们又是大量使用BN，所以成个back－prop的graph破坏了pytorch grad engine。解决方案是写一个cross gpu的autograd function来handle。

作者：张航
链接：https://www.zhihu.com/question/59321480/answer/198077371
来源：知乎
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
```

Could my problem also be related to the third point? For "写一个cross gpu的autograd function", do you mean [this function](https://github.com/zhanghang1989/PyTorch-Encoding/blob/d4e195538a5cac3a33b9cbde28d54c8ed78a3f91/encoding/parallel.py#L26)?

**Comment by zhanghang1989** [comment]
The cross gpu communication is already handled in current implementation.

**Comment by xiongzhanblake** [comment]
Hi @zhanghang1989 @jianchao-li ,
I have the same problem that program gets stuck at [this line](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/experiments/segmentation/train.py#L108). 
I used pytorch 0.4.1, cuda 8.0, gcc 4.9.4 and python 3.6.
As Zhanghang1989 mentioned in [post](https://github.com/zhanghang1989/PyTorch-Encoding/issues/117) 

> The package is no longer compatible with PyTorch 0.4.1, due to the rapid PyTorch updates. 

May I ask where could I find Pytorch 0.5.0 as suggested?

**Comment by jianchao-li** [comment]
I found [this reply](https://github.com/facebookresearch/Detectron/issues/219#issuecomment-400688733) which seems relevant. I am actually using 4 Tesla-V100-PCIE-16GB with CUDA 9.0. I am trying to update the driver now and see whether it will fix the problem.

By the way, I also met this problem with `gluon-cv`. So I think this is not the problem of the codes but of some other stuff like the environment.

**Comment by jianchao-li** [comment]
Hello, @zhanghang1989.

I found something possibly relevant to the deadlock problem using [`p2pBandwidthLatencyTest`](https://github.com/zchee/cuda-sample/blob/master/1_Utilities/p2pBandwidthLatencyTest/p2pBandwidthLatencyTest.cu). Running this test revealed that some GPUs in my machine cannot talk to each other.

```bash
Device: 0, GeForce GTX 1080 Ti, pciBusID: 2, pciDeviceID: 0, pciDomainID:0
Device: 1, GeForce GTX 1080 Ti, pciBusID: 3, pciDeviceID: 0, pciDomainID:0
Device: 2, GeForce GTX 1080 Ti, pciBusID: 82, pciDeviceID: 0, pciDomainID:0
Device: 3, GeForce GTX 1080 Ti, pciBusID: 83, pciDeviceID: 0, pciDomainID:0
Device=0 CAN Access Peer Device=1
Device=0 CANNOT Access Peer Device=2
Device=0 CANNOT Access Peer Device=3
Device=1 CAN Access Peer Device=0
Device=1 CANNOT Access Peer Device=2
Device=1 CANNOT Access Peer Device=3
Device=2 CANNOT Access Peer Device=0
Device=2 CANNOT Access Peer Device=1
Device=2 CAN Access Peer Device=3
Device=3 CANNOT Access Peer Device=0
Device=3 CANNOT Access Peer Device=1
Device=3 CAN Access Peer Device=2
```

**Comment by xiongzhanblake** [comment]
Hi, @jianchao-li.
I tried p2pBandwidthLatencyTest  you suggested, and found my GPUS can communicate to each other. Thus, you might have to figure out this problem first!
Then, I found out the reason for deadlock in multi-gpu training. It turns out to be [syncbn](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/nn/syncbn.py) does not check the error of gpu-memory runing-out @zhanghang1989 . The proof is that when I changed encoding.nn.BatchNorm2d back to original torch.nn.BatchNorm2d, training process detected gpu memory run-out error. And when I used back encoding.nn.BatchNorm2d, no error poped up but a deadlock happened instead. Finally, I  chose a smaller batch-size, no deadlock occurred any more.
Now, I can train my network on cityscapes with encoding pretty well!

I Hope my solution can help you as well! Cheers!

**Comment by jianchao-li** [comment]
Hi, @xiongzhanblake. Thanks for the update. I will try to run it on 4 GPUs using smaller batch size and see whether I will be of any luch 😂

**Comment by jianchao-li** [comment]
Still no luck. Stuck very soon...

**Comment by xiongzhanblake** [comment]
Have you worked out GPU communication issue?

**Comment by jianchao-li** [comment]
Hi, @zhanghang1989, @xiongzhanblake.

> I found [this reply](https://github.com/facebookresearch/Detectron/issues/219#issuecomment-400688733) which seems relevant. I actually came across this problem when using 4 Tesla-V100-PCIE-16GB GPUs with CUDA 9.0. And I also met it when using 4 GeForce-GTX-1080-Ti GPUs.
> 
> By the way, I also met the probably deadlock problem with multi-gpu training of `gluon-cv`. So I think this problem may not be due to codes but some other factors.

As said in the above comment, I switched from `PyTorch-Encoding` to `gluon-cv` for training PSPNet using `SyncBatchNorm`. And the training codes of `gluon-cv` was also stuck. After several days of investigation, I finally found a way to get rid of this problem.

I was running the experiment in a GPU cluster, which is kind of like a cloud environment in which you can set up a virtual machine with specified number of resources (like CPUs, GPUs, memory etc.). And the stuck problem happened when I was using 4 GPUs while using 2 GPUs worked fine. Previously, I set the number of CPUs to be 5 and memory to be 10 GB and it was stuck for 4 GPUs. On this morning I changed it the number of CPUs and memory to be 8 and 20 GB respectively and it worked! Now I am happily training PSPNet on 4 GPUs with `SyncBatchNorm` in `gluon-cv`.

I think the stuck should be due to the resource settings. However, I still have no concrete idea about how it interferes with the training and leads to the stuck. But as the problem is resolved for now, I will close this issue.

**Comment by thienan092** [comment]
I got the same problem. It's still doesn't help even with the number of vCPUs into 24 and vRAM into 128. I tried to reduce the size of the model and it stuck in the middle of the epoch. I believe that increasing _num_workers_ of **DataLoader** makes it worse.
